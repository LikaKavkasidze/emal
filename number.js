/// Emal – Acquire and process scientific data with ease
/// number.js – Parser of arbitrary precision numbers.
///
/// An Emal number is stored as a large integer (BigInt) with decimal
/// place position.
/// Parsing is based on a state machine with states corresponding to
/// parts of the number.

import { StateMachine } from "./stateMachine.js";
import { isDigit, isSign, isComma } from "./matchers.js";

// Default decimal count for `toString`
const ENUMBER_DEFAULT_DC = 2;
// Default extra decimals computed for division
const ENUMBER_DEFAULT_DIV_OFFSET = BigInt(ENUMBER_DEFAULT_DC + 2);

// Define a start state which defines a local variable and simply returns.
// This declared local variable is used to construct tokens generically, so 
// that parsing of the number and exponent uses the same states.
function start(charCode) {
	this.isExponent = false;
	return this.sign(charCode);
}

function sign(charCode) {
	if(!isSign(charCode)) {
		return this.startIntPart(charCode);
	}

	this.token({}, `${this.isExponent ? "eS" : "s"}ign`);
	this.eat(charCode);

	return startIntPart;
}

function startIntPart(charCode) {
	if(!isDigit(charCode)) {
		return this.nok(charCode);
	}

	this.token({}, `${this.isExponent ? "eI" : "i"}ntPart`);
	this.eat(charCode);

	return intPart;
}

function intPart(charCode) {
	if(charCode === 69 || charCode === 101) {
		this.isExponent = true;
		return sign;
	// Exponents cannot have a floating part
	} else if(isComma(charCode) && !this.isExponent) {
		this.token({}, "floatPart");
		return floatPart;
	} else if(isDigit(charCode)) {
		this.eat(charCode);
		return intPart;
	}

	return this.end(charCode);
}

function floatPart(charCode) {
	if(charCode === 69 || charCode === 101) {
		this.isExponent = true;
		return sign;
	} else if(isDigit(charCode)) {
		this.eat(charCode);
		return floatPart;
	}

	return this.end(charCode);
}

// Used as a transformer on token values which are strings
function toNumber(token) {
	return BigInt(token.value);
}

function toValue(token) {
	return token.value;
}

export const numberSm = new StateMachine("start",
	{
		start,
		sign,
		startIntPart,
		intPart,
		floatPart,
	},
	{
		transformers: {
			sign: toValue,
			intPart: toValue,
			floatPart: toValue,
			eSign: toValue,
			eIntPart: toNumber,
		},
		type: "parallel",
	},
);

function BIPow(n, p) {
	let o = n;

	for(; p > 1; p--) {
		o *= n;
	}

	return o;
}

function preprocessOp(a, b) {
	let aInt = a.rawInt;
	let bInt = b.rawInt;

	let aLen = a.decimalPlace;
	let bLen = b.decimalPlace;

	if(aLen > bLen) {
		const delta = aLen - bLen;
		bInt *= BIPow(10n, delta);
		bLen += delta;
	} else if(bLen > aLen) {
		const delta = bLen - aLen;
		aInt *= BIPow(10n, delta);
		aLen += delta;
	}

	return {
		aInt,
		bInt,
		oLen: aLen,
	};
}

export class EmalNumber {
	constructor(rawInt, decimalPlace) {
		this.rawInt = rawInt;
		this.decimalPlace = decimalPlace;
	}

	static fromTokens(userTokens) {
		// Use defaults because some parts of the number are optional
		const defaultTokens = {
			sign: "+",
			floatPart: "0",
			eSign: "+",
			eIntPart: "0",
		};

		const tokens = Object.assign({}, defaultTokens, userTokens);

		const rawInt = BigInt(`${tokens.sign}${tokens.intPart}${tokens.floatPart}`);
		const exponent = BigInt(`${tokens.eSign}${tokens.eIntPart}`);
		const decimalPlace = BigInt(tokens.floatPart.length) - exponent;

		return new EmalNumber(rawInt, decimalPlace);
	}

	static fromString(numberStr) {
		const runnerOutput = numberSm.run(numberStr).transform();
		return EmalNumber.fromTokens(runnerOutput.tokens);
	}

	toString(decimalCount) {
		// Dynamic default argument
		if(typeof decimalCount !== "number") decimalCount = ENUMBER_DEFAULT_DC;

		const { rawInt, decimalPlace } = this;

		const intStr = rawInt.toString(10);
		const exponent = BigInt(intStr.length) - decimalPlace - 1n;

		const intPart = intStr.substr(0, 1);
		const floatPart = intStr.substr(1, decimalCount);

		return `${intPart},${floatPart}${exponent !== 0n ? "e".concat(exponent) : ""}`
	}

	static add(a, b) {
		const { aInt, bInt, oLen } = preprocessOp(a, b);
		return new EmalNumber(aInt + bInt, oLen);
	}

	static sub(a, b) {
		const { aInt, bInt, oLen } = preprocessOp(a, b);
		return new EmalNumber(aInt - bInt, oLen);
	}

	static mul(a, b) {
		const { aInt, bInt, oLen } = preprocessOp(a, b);
		return new EmalNumber(aInt * bInt, oLen * 2n);
	}

	static div(a, b) {
		const { aInt, bInt, oLen } = preprocessOp(a, b);
		// Dividing arbitrary precision numbers cannot be done
		// with infinite precision, here the number is cut to
		// ENUMBER_DEFAULT_DIV_OFFSET digits after decimal point.
		const offset = ENUMBER_DEFAULT_DIV_OFFSET + oLen;
		return new EmalNumber(aInt * BIPow(10n, offset) / bInt, offset);
	}
}
