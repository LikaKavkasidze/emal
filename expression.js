/// Emal – Acquire and process scientific data with ease
/// expression.js – Calculator for generic expressions.
///
/// An expression is a string representing operations to be done.
/// Example: 0.75 + 2 * log(x)
/// Expressions are represented as a sequence of tokens, in Reverse
/// Polish Notation (RPN).

import { isWhitespace, isDigit, isSign, isOpeningBracket, isClosingBracket, isCommaSeparator } from "./matchers.js";
import { EmalNumber, numberSm } from "./number.js";
import { StateMachine } from "./stateMachine.js";

const EEXPR_OPERATORS = {
	"+": {
		priority: 1,
		methodName: "add",
	},
	"-": {
		priority: 1,
		methodName: "sub",
	},
	"*": {
		priority: 2,
		methodName: "mul",
	},
	"/": {
		priority: 2,
		methodName: "div",
	},
};

const EEXPR_FUNCTIONS = {
	"log": {
		argumentCount: 1,
		methodName: "log10",
	},
	"max": {
		argumentCount: 2,
		methodName: "max",
	},
};

const EEXPR_OPERATOR_CHARS = Object.keys(EEXPR_OPERATORS).map(c => c.charCodeAt(0));
const EEXPR_FUNCTION_NAMES = Object.keys(EEXPR_FUNCTIONS);

function isOperator(charCode) {
	return EEXPR_OPERATOR_CHARS.includes(charCode);
}

function exprToken(charCode) {
	if(isDigit(charCode) || isSign(charCode)) {
		const hasDelegated = this.delegate(numberSm, { type: "number" });

		if(hasDelegated) {
			return exprToken;
		}
	}

	const opCond = isOperator(charCode);
	const csCond = isCommaSeparator(charCode);
	const obCond = isOpeningBracket(charCode);
	const cbCond = isClosingBracket(charCode);

	// Operators, brackets and commas are single-eaten under a token
	if(opCond || csCond || obCond || cbCond) {
		if(opCond) {
			this.token({
				type: "operator",
			});
		} else if(csCond) {
			this.token({
				type: "comma",
			});
		} else {
			this.token({
				type: "bracket",
				bracketType: cbCond ? "closing" : "opening",
			});
		}

		this.eat(charCode);
		return exprToken;
	}

	if(isWhitespace(charCode)) {
		return exprToken;
	}

	if(isNaN(charCode)) {
		return this.end(charCode);
	}

	// By default, token is expected to be a variable and will later be
	// mutated if the name matches a function name.
	this.token({ type: "variable" });
	return this.variableOrFunction(charCode);
}

function variableOrFunction(charCode) {
	if(isWhitespace(charCode) || isOperator(charCode) || isOpeningBracket(charCode) || isClosingBracket(charCode) || isCommaSeparator(charCode)) {
		// Mutate token to function if a function name is found
		if(EEXPR_FUNCTION_NAMES.includes(this.eaten())) {
			this.mutate({ type: "function" });
		}

		return this.exprToken(charCode);
	}

	if(isNaN(charCode)) {
		// Mutate token to function if a function name is found
		if(EEXPR_FUNCTION_NAMES.includes(this.eaten())) {
			this.mutate({ type: "function" });
		}

		return this.end(charCode);
	}

	this.eat(charCode);
	return variableOrFunction;
}

const expressionSm = new StateMachine("exprToken", {
	exprToken,
	variableOrFunction,
}, {
	type: "sequential"
});

export class Expression {
	constructor(expressionStr) {
		const runnerOutput = expressionSm.run(expressionStr);

		// State machine transformers cannot be used for sequential parsing
		const tokens = runnerOutput.tokens
			.map(tok => {
				if(tok.type === "number") {
					tok.children = EmalNumber.fromTokens(tok.children);
				}

				return tok;
			});

		// After the lexer, based on a sequential state machine, tokens needs
		// to be ordered in RPN. This code is based on Dijkstra's shunting
		// yard algorithm.
		const rpnOutput = tokens.reduce((acc, cur) => {
			switch(cur.type) {
				case "number":
				case "variable":
					acc.tokens.push(cur);
					break;

				case "function":
					acc.stack.push(cur);
					break;

				case "operator":
					const currentPriority = EEXPR_OPERATORS[cur.value].priority;
					let op = acc.stack.pop();

					while(op && op.type !== "bracket" && EEXPR_OPERATORS[op.value].priority > currentPriority) {
						acc.tokens.push(op);
						op = acc.stack.pop();
					}

					// Push back last popped operator
					if(op) acc.stack.push(op);
					acc.stack.push(cur);
					break;

				case "bracket":
				case "comma":
					if(cur.bracketType === "opening") {
						acc.stack.push(cur);
					} else {
						let op = acc.stack.pop();

						while(op && op.bracketType !== "opening") {
							acc.tokens.push(op);
							op = acc.stack.pop();
						}

						// Do not push anything back: brackets get discarded
					}
					break;
			}

			return acc;
		}, { tokens: [], stack: [] });

		// Fully pop remaining stack
		rpnOutput.stack.reverse().forEach(op => rpnOutput.tokens.push(op));

		this.tokens = rpnOutput.tokens;
	}

	// Evaluating an expression on a set of variables gives the result
	// of the operations represented by the tokens, replacing all variables
	// by their values.
	evaluate(variables) {
		return this.tokens.reduce((acc, tok) => {
			switch(tok.type) {
				case "number":
					acc.push(tok.children);
					break;
				
				case "variable":
					const value = variables[tok.value];
					acc.push(value);
					break;

				case "operator":
					const opMethodName = EEXPR_OPERATORS[tok.value].methodName;
					const opArgs = acc.splice(-2, 2);
					// Assume all arguments are of the same type
					const opMethod = opArgs[0].constructor[opMethodName];

					acc.push(opMethod.apply(null, opArgs));
					break;

				case "function":
					const { methodName, argumentCount } = EEXPR_FUNCTIONS[tok.value];
					const args = acc.splice(-argumentCount, argumentCount);
					const method = args[0].constructor[methodName];

					acc.push(method.apply(null, args));
					break;
			}
	
			return acc;
		}, [])[0];
	}
};
