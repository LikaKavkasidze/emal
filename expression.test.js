import { Expression } from "./expression.js";
import { EmalNumber } from "./number.js";

class MockNumber {
	constructor(inner) {
		this.inner = inner;
	}

	static add(a, b) {
		return new MockNumber(a.inner + b.inner);
	}

	static sub(a, b) {
		return new MockNumber(a.inner - b.inner);
	}

	static mul(a, b) {
		return new MockNumber(a.inner * b.inner);
	}

	static div(a, b) {
		return new MockNumber(a.inner / b.inner);
	}

	static log10(a) {
		return new MockNumber(Math.log10(a.inner));
	}

	static max(a, b) {
		if(a.inner >= b.inner) {
			return new MockNumber(a.inner);
		} else {
			return new MockNumber(b.inner);
		}
	}
}

test("tokenizes simple expression", () => {
	const expr = new Expression("a + b");
	expect(expr.tokens).toMatchSnapshot();
});

test("tokenizes function", () => {
	const expr = new Expression("log(cH3O)");
	expect(expr.tokens).toMatchSnapshot();
});

test("executes simple expression", () => {
	const expr = new Expression("a + b");

	const a = new MockNumber(4.55);
	const b = new MockNumber(1.22);

	expect(expr.evaluate({ a, b }).inner).toEqual(5.77);
});

test("executes function", () => {
	const expr = new Expression("log(cH3O)");
	const cH3O = new MockNumber(1.22);

	expect(expr.evaluate({ cH3O }).inner).toEqual(Math.log10(1.22));
});

test("executes function with multiple arguments", () => {
	const expr = new Expression("max(a; b)");

	const a = new MockNumber(4.55);
	const b = new MockNumber(1.22);

	expect(expr.evaluate({ a, b }).inner).toEqual(4.55);
});

test("tokenizes in RPN", () => {
	const expr = new Expression("a + b * a + (a - b)");
	expect(expr.tokens).toMatchSnapshot();
});

test("delegates number parsing", () => {
	const expr = new Expression("a * -2.5e-2");
	expect(expr.tokens).toMatchSnapshot();
});

test("supports Emal numbers", () => {
	const a = EmalNumber.fromString("1.52");
	const b = EmalNumber.fromString("5e-1");

	const expr = new Expression("(a + b) / 2");
	const res = expr.evaluate({ a, b });

	expect(res.toString()).toEqual("1,01");
});

test("distinguishes number comma and argument separator", () => {
	const expr = new Expression("2,5 * max(a, b)");
	expect(expr.tokens).toMatchSnapshot();
});
