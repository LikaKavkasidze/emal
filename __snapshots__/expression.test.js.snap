// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delegates number parsing 1`] = `
[
  {
    "type": "variable",
    "value": "a",
  },
  {
    "children": EmalNumber {
      "decimalPlace": 3n,
      "rawInt": -25n,
    },
    "type": "number",
    "value": "",
  },
  {
    "type": "operator",
    "value": "*",
  },
]
`;

exports[`distinguishes number comma and argument separator 1`] = `
[
  {
    "children": EmalNumber {
      "decimalPlace": 1n,
      "rawInt": 25n,
    },
    "type": "number",
    "value": "",
  },
  {
    "type": "variable",
    "value": "a",
  },
  {
    "type": "variable",
    "value": "b",
  },
  {
    "type": "function",
    "value": "max",
  },
  {
    "type": "operator",
    "value": "*",
  },
]
`;

exports[`tokenizes function 1`] = `
[
  {
    "type": "variable",
    "value": "cH3O",
  },
  {
    "type": "function",
    "value": "log",
  },
]
`;

exports[`tokenizes in RPN 1`] = `
[
  {
    "type": "variable",
    "value": "a",
  },
  {
    "type": "variable",
    "value": "b",
  },
  {
    "type": "variable",
    "value": "a",
  },
  {
    "type": "operator",
    "value": "*",
  },
  {
    "type": "variable",
    "value": "a",
  },
  {
    "type": "variable",
    "value": "b",
  },
  {
    "type": "operator",
    "value": "-",
  },
  {
    "type": "operator",
    "value": "+",
  },
  {
    "type": "operator",
    "value": "+",
  },
]
`;

exports[`tokenizes simple expression 1`] = `
[
  {
    "type": "variable",
    "value": "a",
  },
  {
    "type": "variable",
    "value": "b",
  },
  {
    "type": "operator",
    "value": "+",
  },
]
`;
